test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/ViewDataDictionaryTest.cs

```diff
@@ -311,30 +390,64 @@ public void CopyConstructors_CopyModelMetadata_ForTypeObject()
-        [InlineData(typeof(IEnumerable<string>), new object[] { "1", "2", "3", }, typeof(object[]))]
-        [InlineData(typeof(List<string>), new object[] { 1, 2, 3, }, typeof(object[]))]
-        public void CopyConstructors_OverrideSourceMetadata_IfModelNonNull(
+        [InlineData(typeof(IEnumerable<string>), new object[] { "1", "2", "3" }, typeof(object[]))]
+        [InlineData(typeof(List<string>), new object[] { 1, 2, 3 }, typeof(object[]))]
+        public void CopyConstructors_UpdateModelMetadata_IfModelIncompatibleWithSourceMetadata(
             Type sourceType,
-            object instance,
+            object model,
             Type expectedType)
         {
             // Arrange
             var metadataProvider = new EmptyModelMetadataProvider();
-            var source = new ViewDataDictionary(metadataProvider);
+            var source = new ViewDataDictionary(metadataProvider)
+            {
+                ModelExplorer = metadataProvider.GetModelExplorerForType(sourceType, model: null),
+            };
+            var sourceMetadata = source.ModelMetadata;
 
             // Act
             var viewData1 = new ViewDataDictionary(source)
             {
-                Model = instance,
+                Model = model,
             };
-            var viewData2 = new ViewDataDictionary(source, model: instance);
+            var viewData2 = new ViewDataDictionary(source, model: model);
 
             // Assert
-            Assert.NotNull(viewData1.ModelMetadata);
+            Assert.NotSame(source.ModelExplorer, viewData1.ModelExplorer);
+            Assert.NotSame(source.ModelMetadata, viewData1.ModelMetadata);
+            Assert.Equal(expectedType, viewData1.ModelMetadata.ModelType);
+
+            Assert.NotSame(source.ModelExplorer, viewData2.ModelExplorer);
+            Assert.NotSame(source.ModelMetadata, viewData2.ModelMetadata);
             Assert.Equal(expectedType, viewData1.ModelMetadata.ModelType);
+        }
+
+        [Theory]
+        [InlineData(typeof(int), 23)]
+        [InlineData(typeof(string), "test string")]
+        [InlineData(typeof(IEnumerable<string>), new string[] { "1", "2", "3" })]
+        public void CopyConstructors_PreservesSourceMetadata_IfModelNonNull(Type sourceType, object model)
+        {
+            // Arrange
+            var metadataProvider = new EmptyModelMetadataProvider();
+            var source = new ViewDataDictionary(metadataProvider)
+            {
+                ModelExplorer = metadataProvider.GetModelExplorerForType(sourceType, model: null),
+            };
+
+            // Act
+            var viewData1 = new ViewDataDictionary(source)
+            {
+                Model = model,
+            };
+            var viewData2 = new ViewDataDictionary(source, model: model);
+
+            // Assert
+            Assert.NotSame(source.ModelExplorer, viewData1.ModelExplorer);
+            Assert.Same(source.ModelMetadata, viewData1.ModelMetadata);
```

Comment:
This assertion fails for the `typeof(IEnumerable<string>), new string[] { "1", "2", "3" })` case w/o the `src` changes. `viewData1.ModelMetadata.ModelType` is `typeof(string[])` and the same for `viewData2`. Not exactly _wrong_ in this case (no property metadata was lost) but property metadata would be lost if it were available in `source.ModelMetadata`.